File: Breadth-First Search/Matrix pattern/LET_no994_Rotting Oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        freshes = 0
        minutes = 0
        queue = collections.deque([])
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    queue.append((i, j))
                elif grid[i][j] == 1:
                    freshes += 1
                    
        while queue:
            for _ in range(len(queue)):
                i, j = queue.popleft()
                for di, dj in [(1, 0), (0, 1), (-1, 0), (0, -1)]:
                    i2, j2 = i + di, j + dj
                    if 0 <= i2 < m and 0 <= j2 < n and grid[i2][j2] == 1:
                        freshes -= 1
                        grid[i2][j2] = 2
                        queue.append((i2, j2))
            if queue:
                minutes += 1
        
        return minutes if freshes == 0 else -1

File: Breadth-First Search/Matrix pattern/LET_no994_Rotting Oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        queue<pair<int, int>> q;
        int m = grid.size(), n = grid[0].size(), fresh = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 2)
                    q.push({i, j});
                if (grid[i][j] == 1)
                    fresh++;
            }
        }
        
        int res = 0, i2, j2;
        int shifts[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        while (!q.empty()) {
            bool nxt = false;
            for (int i = q.size(); i > 0; i--) {
                auto p = q.front();
                q.pop();
                for (int j = 0; j < 4; j++) {
                    i2 = p.first + shifts[j][0];
                    j2 = p.second + shifts[j][1];
                    if (0 <= i2 && i2 < m && 0 <= j2 && j2 < n && grid[i2][j2] == 1) {
                        nxt = true;
                        grid[i2][j2] = 2;
                        fresh--;
                        q.push({i2, j2});
                    }
                }
            }
            res += nxt;
        }
        
        return fresh == 0 ? res : -1;
    }
};